_posts/2020/2020-02-11-unit-testing-in-react.md

```diff
@@ -0,0 +1,251 @@
+
+
+const ProductList = () => {
+  const [products, setProducts] = useState([])
+  const [page, setPage] = useState(1)
+  const loadProducts = async () => {
+    const response = await fetchProducts(page)
+    if(response) {
+      const newProducts = response.products
+      setProducts(products.concat(newProducts))
+      setPage(page + 1)
+    }
+  }
+  return <div>
+    <button onClick={loadProducts}>
+       Load more!
+    </button>
+    {products.map(product => <Product {...product} />)}
+  </div>
+}
+
+export default ProductList;
+```
+
+Before going on to how to test it, I must say I've prepared a sample repo for you to try out all the code in this post.
+Just check it out on Github [here](https://github.com/DanielHara/unit-testing-react){:target="_blank"}.
+
+## Let's test it!
+
+Now comes the question, how would you test the component `ProductList`?
+
+Here at ePages we use extensively the libraries `react-testing-library` and `jest` to write our unit tests.
+The former contains lots of useful features, like querying for HTML elements and firing all the events the user can trigger, while the
+latter provides us with a thorough mocking and assertion library.
+As soon as I'm done writing the first lines of the tests, I imagine that my first test case would be to try to find a button
+to load some products:
+
+```javascript
+import { fireEvent, render } from '@testing-library/react'
+
+it('should render a button to load products', () => {
+  const { getByText } = render(<ProductList />)
+
+  const loadMoreButton = getByText('Load more!')
+  fireEvent.click(loadMoreButton)
+})
+```
+
+But, now, what I should expect? I do not know what some external API will return me. Maybe it is temporarily down.
+There comes an important notion of testing: you must decouple your tests from other dependencies as much as possible. After all, are you testing
+just `ProductList` or also an external API?
+
+## Mock it
+
+`jest` offers you the possibility of mocking all kinds of functions, including their return values and assert how many times they were called,
+and with which arguments they were called. It also offers a great deal of synthatic sugar to make your tests look pretty and for you to admire them
+after they've been written.
+
+Maybe we could mock `fetchProducts` to take a look on whether it was called with the right arguments, and also mock the return value,
+to assert whether the `ProductList` is rendering them correctly. This way our tests becomes a lot more predictable.
+
+However... Can you do it? It's now difficult to mock `fetchProducts`, because you'd have to override the `ProductList.jsx` file.
+There comes another hint: try to inject your external dependencies such that you can mock them! You could pass `fetchProducts` as a prop instead:
+
+```javascript
+const defaultFetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
+  .then(response => response.json())
+  .catch((error) => console.log(error))
```

Comment:
For brevity in this blog post, I suggest to leave out the `.catch()`.

Reply:
done